source/Target/SwiftLanguageRuntime.cpp

```diff
@@ -689,26 +691,54 @@ bool SwiftLanguageRuntime::IsSwiftMangledName(const char *name) {
   return swift::Demangle::isSwiftSymbol(name);
 }
 
-std::string SwiftLanguageRuntime::DemangleSymbolAsString (const char *symbol,
-                                                          bool simplified) {
-  if (simplified) {
-    swift::Demangle::DemangleOptions options(swift::Demangle::DemangleOptions::
-                                             SimplifiedUIDemangleOptions());
-    return swift::Demangle::demangleSymbolAsString(
-        symbol, strlen(symbol), options);
-  } else
-    return swift::Demangle::demangleSymbolAsString(symbol, strlen(symbol));
-}
+std::string
+SwiftLanguageRuntime::DemangleSymbolAsString(StringRef symbol, bool simplified,
+                                             const SymbolContext *sc) {
+  swift::Demangle::DemangleOptions options;
+  llvm::DenseMap<std::pair<uint64_t, uint64_t>, StringRef> dict;
+  if (simplified)
+    options = swift::Demangle::DemangleOptions::SimplifiedUIDemangleOptions();
+
+  if (sc) {
+    bool can_create = true;
+    // This terrifying cast avoids having too many differences with llvm.org.
+    if (Block *block = const_cast<SymbolContext *>(sc)->GetFunctionBlock())
+      if (VariableListSP var_list = block->GetBlockVariableList(can_create))
+        for (unsigned i = 0; i < var_list->GetSize(); ++i) {
```

Comment:
Instead of having this lexing process here, can we move this logic to `swift::Demangle`, and maybe add some unit tests for it?

Reply:
I don't think that makes sense. That format is debug-info-specific and has nothing to do with (de)mangling. It's the naming format we use for the type metadata variables in DWARF.

Reply:
If you feel lldb is the right place where this function should live, I think we should at least factor this out so we can test it in isolation :)

Reply:
I factored out common code between SwiftASTContext and SwiftLanguageRuntime into a new `SwiftLanguageRuntime::GetArchetypeNamesForFunction()`.